### PR TITLE
sixth assignment: rearranging cars

### PR DESCRIPTION
**Pull request by yuvdanilova**
## Task 6, assignment6.py

**Description:**
We are given two arrays of the same size which describe positions of cars on parking lots, initial and desired. There is exactly one empty slot, denoted by 0. We should rearrange cars in the given order, though the only possible action at one moment is to move one particular car to an empty slot.

### Solution

This task could be approached as a simple sorting problem with specific swap and comparison functions.

**Swapping**:
Given objects [__a, b, c__] at indices [__i, j, k__] we can swap __a__ and __b__ by swapping these with __c__ in three operations:

1) __a <-> c__, swap elements at __i__ and __k__, get [__c, b, a__]
2) __b <-> c__, swap elements at __i__ and __j__, get [__b, c, a__]
2) __c <-> a__, swap elements at __j__ and __k__, get [__b, a, c__]

Thus if we know where the empty slot is located, we can swap any two cars in O(1), and the empty slot remains on the same place.

**Sorting order**:
Using a target list, we can create a dictionary object where keys are our cars and values are their positions in the final arrangement. If we use a comparison-based sorting algorithm, we obtain the right order by comparing those values.

**Sorting algorithm**:
I decided to use QuickSort as a sorting algorithm. I also thought of HeapSort, which is O(nlogn)  worst case, but its disadvantage is that it usually makes more swaps, and in this task we made the swapping procedure more complicated.

Modifications: first I move the empty slot to the end of the array, then sort N-1 elements, then move the empty slot to the final position.

### Asymptotic analysis

**Time complexity**:
O(n) for creating a dict
O(n) for moving the empty slot
O(nlogn) - QuickSort average case, O(n^2) - QuickSort worst case
Total: __O(nlogn)__ - average case, __O(n^2)__ - worst case

**Space complexity**:
__O(n)__ for the order dictionary

### CODE REVIEW

**Review by ReemaJiries — APPROVED** [review]
Your Code is very good.

File: assignment6.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
+from statistics import median
+import unittest
+import numpy as np
+
+class RearrangingArray():
+
+	def __init__(self, arrangement, verbose=False):
```
Comment (galeal2) [inline assignment6.py:8]: I was thinking maybe would be useful to explain a little bit the arguments from the 'init'

File: assignment6.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+		""" Swaps elements on indices i and j using an element on index k """
+		if self.verbose:
+			print(i, k)
+			print(i, j)
+			print(j, k)
+		self.arrangement[i],self.arrangement[k] = self.arrangement[k],self.arrangement[i]
+		self.arrangement[i],self.arrangement[j] = self.arrangement[j],self.arrangement[i]
+		self.arrangement[j],self.arrangement[k] = self.arrangement[k],self.arrangement[j]
+
+	def _move_swap_el(self, j):
```
Comment (galeal2) [inline assignment6.py:25]: 😁it took me a little bit to get that 'el' comes from 'elements' for me would be easier 'elm' maybe
Just an idea 🙂

File: assignment6.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	def test_random(self):
+
+		for j in range(1, 10001, 10):
+			np.random.seed(j)
+			init_order = list(np.random.permutation(np.arange(j)))
+			np.random.seed(j * 2)
+			new_order = list(np.random.permutation(np.arange(j)))
+			rarr = RearrangingArray(init_order)
+			rarr.sort(new_order)
+			self.assertEqual(rarr.arrangement, new_order)
```
Comment (galeal2) [inline assignment6.py:113]: Like your testing 👌🏻

**Review by galeal2 — COMMENTED** [review]
Nice 😉

File: assignment6.py (hunk 2 above)
Comment (yuvdanilova) [reply assignment6.py:25]: Oh, I'd be confused with 'elm' :)
Maybe 'elem' could work for everyone :)

File: assignment6.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python3
```
Comment (OlgaVorokh) [inline assignment6.py:1]: JFYI, try to use 4 spaces, not tab in your work!

File: assignment6.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+		return order_dict
+
+	def _partition(self, l, r):
+		""" Partition function for QuickSort sorting algorithm """
+		pivot = median([
+			self.order_dict[self.arrangement[l]], 
+			self.order_dict[self.arrangement[(l+r)//2]], 
+			self.order_dict[self.arrangement[r]]]
+		)
+		while True:
```
Comment (OlgaVorokh) [inline assignment6.py:62]: : D 
I also write while in that way, but as I know, this way of code style coders avoiding in job)

**Review by OlgaVorokh — APPROVED** [review]
Great job, Julia!

File: assignment6.py (hunk 5 above)
Comment (yuvdanilova) [reply assignment6.py:62]: @OlgaVorokh ok, I'll try to avoid that too!

**Review by ariava — APPROVED** [review]
Nice work! Please try to incorporate your comments in the code to explain it to the reader.

**Pull request merged**